lib/BoundaryValueDiffEqFIRK/src/firk.jl

```diff
@@ -514,40 +535,54 @@ function __construct_nlproblem(
     (; stage) = cache
     N = length(cache.mesh)
 
-    lossₚ = iip ? ((du, u) -> loss(du, u, cache.p)) : (u -> loss(u, cache.p))
-
     resid_collocation = __similar(y, cache.M * (N - 1) * (stage + 1))
 
     resid = vcat(
         @view(cache.bcresid_prototype[1:prod(cache.resid_size[1])]), resid_collocation,
         @view(cache.bcresid_prototype[(prod(cache.resid_size[1]) + 1):end]))
     L = length(cache.bcresid_prototype)
 
-    sd = if jac_alg.nonbc_diffmode isa AutoSparse
+    diffmode = if jac_alg.diffmode isa AutoSparse
         block_size = cache.M * (stage + 2)
         J_full_band = BandedMatrix(
             Ones{eltype(y)}(L + cache.M * (stage + 1) * (N - 1),
                 cache.M * (stage + 1) * (N - 1) + cache.M),
             (block_size, block_size))
-        __sparsity_detection_alg(__generate_sparse_jacobian_prototype(
-            cache, cache.problem_type,
+        colored_result = __generate_sparse_jacobian_prototype(cache, cache.problem_type,
             @view(cache.bcresid_prototype[1:prod(cache.resid_size[1])]),
             @view(cache.bcresid_prototype[(prod(cache.resid_size[1]) + 1):end]),
-            cache.M, N))
+            cache.M, N, jac_alg.diffmode)
+        partition = ifelse(
+            ADTypes.mode(jac_alg.bc_diffmode) isa ADTypes.ReverseMode, :row, :column)
+        constant_matrix_coloring = ifelse(
+            ADTypes.mode(jac_alg.nonbc_diffmode) isa ADTypes.ReverseMode,
+            row_colors, column_colors)(colored_result)
+        AutoSparse(get_dense_ad(jac_alg.diffmode);
+            sparsity_detector = ADTypes.KnownJacobianSparsityDetector(sparsity_pattern(colored_result)),
+            coloring_algorithm = ConstantColoringAlgorithm{partition}(
+                sparsity_pattern(colored_result), constant_matrix_coloring))
```

Comment:
I don't understand this pattern of first solving the coloring problem and then using a `ConstantColoringAlgorithm` to hold the result. The coloring problem is only solved once during preparation anyway.

Reply:
You mean we don't have to use `sparsity_pattern` to get the coloring of the Jacobian since the colorings will be computed in the preparation?

Reply:
Yes, the preparation includes sparsity pattern detection and coloring, and all of this is transparent to the user. To me it seems like you re-create a new backend object containing the fixed result of coloring, instead of just keeping the provided backend and letting coloring do its thing at preparation time under the hood.

Reply:
If I understand correctly, since we have the sparsity pattern of our sparse Jacobian, we can just use

```julia
AutoSparse(get_dense_ad(jac_alg.diffmode);
    sparsity_detector = ADTypes.KnownJacobianSparsityDetector(J),
    coloring_algorithm = GreedyColoringAlgorithm())
```
to construct our `AutoSparse` backend and let preparation do their work?

Reply:
Yes